sw/device/lib/runtime/epmp.h

```diff
@@ -0,0 +1,550 @@
+ * it is not possible to set these values. The combination R=1 W=1
+ * X=1 has also been reserved by this library and may not be used
+ *
+ * An entry may only be configured with unlocked permissions if the
+ * entry is also configured as OFF.
+ *
+ * Note: permissions may have different meanings when Machine Mode
+ * Lockdown (mseccfg.MML) is set.
+ */
+typedef enum {
+  /**
+   * Unlocked with R=0, W=0 and X=0.
+   *
+   * Note: full access (R=1, W=1 and X=1) in Machine Mode. Only use
+   * in disabled (OFF) entries.
+   */
+  kEpmpPermUnlocked,
+
+  /**
+   * Locked with R=0, W=0 and X=0.
+   */
+  kEpmpPermLockedNoAccess,
+
+  /**
+   * Locked with R=0, W=0 and X=1.
+   */
+  kEpmpPermLockedExecuteOnly,
+
+  /**
+   * Locked with R=1, W=0 and X=0.
+   */
+  kEpmpPermLockedReadOnly,
+
+  /**
+   * Locked with R=1, W=0 and X=1.
+   */
+  kEpmpPermLockedReadExecute,
+
+  /**
+   * Locked with R=1, W=1 and X=0.
+   */
+  kEpmpPermLockedReadWrite,
+} epmp_perm_t;
+
+/**
+ * EPMP region specification.
+ *
+ * Provides the start and end addresses of a particular region. These addresses
+ * are byte-aligned (i.e. they are like regular pointers rather than encoded
+ * addresses).
+ *
+ * The `start` address is inclusive and the `end` address is exclusive.
+ */
+typedef struct epmp_region {
+  uintptr_t start;
+  uintptr_t end;
+} epmp_region_t;
+
+/**
+ * EPMP entry index.
+ */
+typedef enum {
+  kEpmpEntry0,
+  kEpmpEntry1,
+  kEpmpEntry2,
+  kEpmpEntry3,
+  kEpmpEntry4,
+  kEpmpEntry5,
+  kEpmpEntry6,
+  kEpmpEntry7,
+  kEpmpEntry8,
+  kEpmpEntry9,
+  kEpmpEntry10,
+  kEpmpEntry11,
+  kEpmpEntry12,
+  kEpmpEntry13,
+  kEpmpEntry14,
+  kEpmpEntry15,
+} epmp_entry_t;
+
+/**
+ * EPMP entry count.
+ */
+typedef enum {
+  kEpmpEntryCount0,
+  kEpmpEntryCount1,
+  kEpmpEntryCount2,
+  kEpmpEntryCount3,
+  kEpmpEntryCount4,
+  kEpmpEntryCount5,
+  kEpmpEntryCount6,
+  kEpmpEntryCount7,
+  kEpmpEntryCount8,
+  kEpmpEntryCount9,
+  kEpmpEntryCount10,
+  kEpmpEntryCount11,
+  kEpmpEntryCount12,
+  kEpmpEntryCount13,
+  kEpmpEntryCount14,
+  kEpmpEntryCount15,
+  kEpmpEntryCount16,
+} epmp_entry_count_t;
+
+/**
+ * EPMP generic status codes.
+ *
+ * These error codes can be used by any function. However if a function
+ * requires additional status codes, it must define a set of status codes to
+ * be used exclusively by such function.
+ */
+typedef enum {
+  kEpmpOk = 0,
+  kEpmpError,
+  kEpmpBadArg,
+} epmp_result_t;
+
+/**
+ * Initialize the EPMP library.
+ *
+ * The current state of the EPMP CSRs will be read and stored inside
+ * library. This function must be called before using the library. The
+ * library can only be initialized once.
+ *
+ * Note: it may be desirable to check the state is as expected. This
+ * can be done using `epmp_get_state`.
+ */
+epmp_result_t epmp_init(void);
+
+/**
+ * Check that the actual EPMP state matches what the library expects
+ * it to be.
+ */
+epmp_result_t epmp_check(void);
```

Comment:
Maybe this should also return a checksum so that the caller can also verify the state.

Reply:
I think we should revisit the checksum idea as we do the integration. I agree it would be good to have something.